### ISSUE TITLE
Invalide global IDs being requested by github plugin?

### ISSUE THREAD

**Issue by Beanow** [sourcecred/sourcecred#1408]
Trying to load `expressjs/express` a massive error message comes back to do with the graphql mirroring.

For example here: https://drone.teamopen.dev/teamopen-dev/sourcecred-stack-lookup/132/1/3
After a bunch of scrolling through it eventually spits out errors like:
```json
{"type":"NOT_FOUND","path":["owndata_2",73],"locations":[{"line":1,"column":1066}],"message":"Could not resolve to a node with the global id of 'MDg6UmVhY3Rpb241MjU4MzY3Nw=='."}]}

"errors":[{"type":"NOT_FOUND","path":["owndata_2",53],"locations":[{"line":1,"column":1066}],"message":"Could not resolve to a node with the global id of 'MDg6UmVhY3Rpb240MjQyMTcxMA=='."},{"type":"NOT_FOUND","path":["owndata_2",73],"locations":[{"line":1,"column":1066}],"message":"Could not resolve to a node with the global id of 'MDg6UmVhY3Rpb241MjU4MzY3Nw=='."}]}} 
```

Seems like we're requesting IDs that perhaps no longer exist?

**Comment by decentralion** [comment]
cc @wchargin 
I wonder if it could be that a node deletion (#938) happened during your load? What happens if you try again with a clean cache?

**Comment by Beanow** [comment]
Looks like more projects are getting this issue over time: https://drone.teamopen.dev/teamopen-dev/sourcecred-stack-lookup/139/1/4

After https://github.com/sourcecred/sourcecred/pull/1407 is done and validated, I would have to clear the cache on the stack lookup cronjob.

Meanwhile if a cache invalidation can solve it, I could try running locally.

**Comment by nothingismagick** [comment]
Doesn't that happen when users morph into orgs?

>

**Comment by Beanow** [comment]
Possibly. I tried base64 decoding the missing IDs to see their types:
```
  'MDQ6VXNlcjU5MzAzMjM' => '04:User5930323',
  'MDU6SXNzdWU1MDQ2ODUzMDg' => '05:Issue504685308',
  'MDg6UmVhY3Rpb24yNjQ5NDQ2MA' => '08:Reaction26494460',
  'MDg6UmVhY3Rpb24yNjQ5NDQ2MQ' => '08:Reaction26494461',
  'MDg6UmVhY3Rpb24yNjQ5NDQ2Mg' => '08:Reaction26494462',
  'MDg6UmVhY3Rpb24yMTQyNTE5MQ' => '08:Reaction21425191',
  'MDg6UmVhY3Rpb240MjQyMTcxMA' => '08:Reaction42421710',
  'MDg6UmVhY3Rpb241MjU4MzY3Nw' => '08:Reaction52583677',
  'MDEyOklzc3VlQ29tbWVudDQ0NjYwNzM0OQ' => '012:IssueComment446607349',
  'MDE3OlB1bGxSZXF1ZXN0UmV2aWV3MzAzNzQ1NzA' => '017:PullRequestReview30374570'
```

There's a fair number of reactions. I wonder if triggering this is as simple as removing your thumbsup. But also an issue, comment, PR review and a user.

So we may be looking at all of them:
- User into org
- Deleted / moved issue
- Deleted issue comment
- Deleted PR review? (Is that an option?)

**Comment by vsoch** [comment]
Is this using the issues api or something else? One thing to check might be if you are getting PRs (and associated content) back that have since been squashed or similar so they aren’t found. It could be that they are linked from somewhere but don’t actually exist anymore.

**Comment by Beanow** [comment]
The mirror code uses GiHubs "v4" GraphQL API. The full query that causes errors is included in the logs. The errors show up when trying to load these nodes by ID. The IDs must have been found somewhere, as @decentralion suggested, probably in our cache, because it is / was connected to something else that we've already mirrored.

I read the suggestions of #938 and it makes sense to partially invalidate the cache to clean up these deleted connections. After all, it's "just cache" and the GitHub data should be authoritative.
Assuming that is the issue of course.

@decentralion do you know how deleting nodes / edges from the mirror would impact the cred calculations? Is it possible for people to lose cred they received a long time ago if for example an issue is deleted?

**Comment by wchargin** [comment]
You can spelunk through the SQL cache to see how you got references to
these items (in-edges in `nodes` and `connection_entries`), which can
sometimes be helpful for understanding.

We don’t currently have any support for deletions, as you saw in #938,
but it’s certainly true that if you _recalculate_ cred after deleting
some nodes then people will lose cred that they received a long time
ago. There’s no persistent state across cred calculations by definition.

**Comment by wchargin** [comment]
Yes, this is likely a node deletion. We requested that ID because we got
it from somewhere, and we don’t ever fabricate IDs, so the API gave us
an ID that either has never been valid (possible, but we haven’t seen
this before) or was valid but no longer is (known failure mode per
#938). The ID decodes to `08:Reaction42421710`, and reactions are
deleted as soon as the user un-reacts to the post, so this is totally
plausible.

**Comment by Beanow** [comment]
> reactions are deleted as soon as the user un-reacts to the post, so this is totally plausible.

Yikes. No wonder the majority of them are reactions.

> You can spelunk through the SQL cache to see how you got references to these items

If this is helpful to finding a solution I should be able to do that. What would we need to verify beyond what's in #938?

**Comment by wchargin** [comment]
The point of spelunking is to look for context: e.g., the owner of the
missing reaction (if you know who that is) commented in the thread, or
the thread has a “repository owner deleted 1 comment” timeline event in
the UI. The prior on reactions is strong enough that it’s probably not
worth it, and an issue comment may also just be a deletion, but the
others are interesting.

For the user node, I don’t think that this is a case of the user turning
into an organization, because in my experiments [that transformation
usually preserves the database ID][gp], but there is no organization
with ID `MDEyOk9yZ2FuaXphdGlvbjU5MzAzMjM=` (`012:Organization5930323`).

It is sometimes possible to stumble across traces of deleted users based
on their database IDs (e.g., noreply email addresses), but I don’t know
of any way to systematically find these.

I am particularly interested in what happened to the pull request
review, as I do not know of any way to delete these from the UI. One way
that this might happen is if the repository’s permissions became more
restricted, so the node actually exists but is 404ing if you don’t have
the right authorization. Can you comment as to whether this is the case?
Here it would be useful to see which pull request has an is-parent edge
to the pull request review. You probably have that pull request’s URL
stored directly; if not, you at least have its enclosing repository and
its index in the list of pull requests.

A query to get you started:

```sql
SELECT object_id
FROM connections JOIN connection_entries ON connections.rowid = connection_id
WHERE child_id = 'MDE3OlB1bGxSZXF1ZXN0UmV2aWV3MzAzNzQ1NzA';
```

I’d also be interested to know what happened to the issue, though this
might be an [issue transfer][it], which I haven’t looked into.

[gp]: https://github.com/wchargin/ghost-playground/issues/2#issuecomment-435236022
[it]: https://github.com/sourcecred/sourcecred/issues/938#issuecomment-435512120

**Comment by Beanow** [comment]
Got it. I'll have a go at this tomorrow.

**Comment by decentralion** [comment]
> @decentralion do you know how deleting nodes / edges from the mirror would impact the cred calculations? Is it possible for people to lose cred they received a long time ago if for example an issue is deleted?

Yeah, if the nodes go missing, the cred disappears. Based on the spec for the [instance system](https://discourse.sourcecred.io/t/sourcecred-instance-system/244) we want to have the full graph stored with every cred snapshot, so it would be technically feasible to persist the lost nodes for cred attribution purposes. However, it would be weird ("ghost nodes") because we might not have other information needed to apply new heuristics, e.g. we don't have the text of the issue so we can't score it with our new fancy ML model.

Ideally, in the future, open-source development moves to an open-source and content addressable platform. For the present time, I'm OK with the status quo that if issues and threads are deleted, then the cred disappears. In exceptional cases (e.g. abuse) we can look into using the historical graphs to mitigate the issue.
